libmultilabel/model.py

```diff
@@ -1,218 +1,112 @@
-            }
-        """
-        # Eval mode
-        self.network.eval()
-
-        # Transfer to GPU
-        for key in inputs:
-            if isinstance(inputs[key], torch.Tensor):
-                inputs[key] = inputs[key].to(self.device, non_blocking=True)
-
-        # Run forward
-        with torch.no_grad():
-            outputs = self.network(inputs['text'])
-            logits = outputs['logits']
-            batch_label_scores = torch.sigmoid(logits)
-
-        return {
-            'scores': batch_label_scores,
-            'logits': logits,
-            'outputs': outputs,
-        }
-
-    def save(self, epoch, is_best=False):
-        self.network.eval()
-        ckpt = {
-            'epoch': epoch,
-            'run_name': self.config.run_name,
-            'state_dict': self.network.state_dict(),
-            'word_dict': self.word_dict,
-            'classes': self.classes,
-            'optimizer': self.optimizer.state_dict(),
-            'best_metric': self.best_metric,
-        }
-        ckpt_path = os.path.join(self.config.result_dir,
-                                 self.config.run_name, 'model_last.pt')
-        os.makedirs(os.path.dirname(ckpt_path), exist_ok=True)
-        logging.info(f"Save current  model: {ckpt_path}")
-        torch.save(ckpt, ckpt_path)
-        if is_best:
-            best_ckpt_path = ckpt_path.replace('last', 'best')
-            logging.info(f"Save best model ({self.config.val_metric}: {self.best_metric}): {best_ckpt_path}")
-            shutil.copyfile(ckpt_path, best_ckpt_path)
-        self.network.train()
-
-    @staticmethod
-    def load(config, ckpt_path):
-        ckpt = torch.load(ckpt_path)
-        return Model(config, ckpt=ckpt)
-
-    def load_best(self):
-        best_ckpt_path = os.path.join(self.config.result_dir,
-                                      self.config.run_name, 'model_best.pt')
-        best_model = self.load(self.config, best_ckpt_path)
-        self.network = best_model.network
+        return optimizer
+
+    @abstractmethod
+    def shared_step(batch):
+        """Return loss and predicted logits"""
+        return NotImplemented
+
+    def training_step(self, batch, batch_idx):
+        loss, _ = self.shared_step(batch)
+        return loss
+
+    def validation_step(self, batch, batch_idx):
+        loss, pred_logits = self.shared_step(batch)
+        return {'loss': loss.item(),
+                'pred_scores': torch.sigmoid(pred_logits).detach().cpu().numpy(),
+                'target': batch['labels'].detach().cpu().numpy()}
+
+    def validation_epoch_end(self, step_outputs):
+        eval_metric = MultiLabelMetrics(self.config)
+        for step_output in step_outputs:
+            eval_metric.add_values(y_pred=step_output['pred_scores'],
+                                   y_true=step_output['target'])
+        eval_metric.eval()
+        self.log_dict(eval_metric.get_metric_dict())
```

Comment:
Set `use_cache` to False will run `eval()` before getting the metric dict. We add `use_cache` to calculate metrics in single run.
```
self.log_dict(eval_metric.get_metric_dict(use_cache=False))
```

Reply:
Cool.

Reply:
What would the evaluation process of local_rank=1,2,3 like if there are 4 GPUs ?

Reply:
If we are using `ddp`, each process will only use `sample size / GPU size` samples for evaluation (DistibutedRamdomSampler) by default in PL, the metrics may not be consistent across processes.  
We could override the sampler in dataloader only for dev/test loader to allow evaluation use whole dataset (https://github.com/ASUS-AICS/icd10-ai/blob/pl/icd/trainer.py)

Reply:
Note: `use_cache` = `False` as default